icicle_v3/backend/cpu/include/ntt_template.h

```diff
@@ -0,0 +1,264 @@
+    if (s_ntt_domain<S> != nullptr) { return eIcicleError::SUCCESS; }
+    // (2) build the domain
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = new CpuNttDomain<S>();
+
+    bool found_logn = false;
+    S omega = primitive_root;
+    unsigned omegas_count = S::get_omegas_count();
+    for (int i = 0; i < omegas_count; i++) {
+      omega = S::sqr(omega);
+      if (!found_logn) {
+        ++s_ntt_domain<S>->max_log_size;
+        found_logn = omega == S::one();
+        if (found_logn) break;
+      }
+    }
+
+    s_ntt_domain<S>->max_size = (int)pow(2, s_ntt_domain<S>->max_log_size);
+    if (omega != S::one()) {
+      ICICLE_LOG_ERROR << "Primitive root provided to the InitDomain function is not in the subgroup";
+      return eIcicleError::INVALID_ARGUMENT;
+    }
+
+    // calculate twiddles
+    // Note: radix-2 INTT needs ONE in last element (in addition to first element), therefore have n+1 elements
+
+    s_ntt_domain<S>->twiddles = (S*)malloc((s_ntt_domain<S>->max_size + 1) * sizeof(S));
+    generate_twiddles(primitive_root, s_ntt_domain<S>->twiddles, s_ntt_domain<S>->max_size);
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t>
+  eIcicleError cpu_ntt_release_domain(const Device& device)
+  {
+    // release the memory
+    free(s_ntt_domain<S>->twiddles);
+    delete s_ntt_domain<S>;
+    s_ntt_domain<S> = nullptr;
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename E = scalar_t>
+  eIcicleError bit_reverse(int size, int logn, E* output, int batch_size)
+  {
+    ICICLE_LOG_DEBUG << "BIT REVERSE";
+    int total_size = size * batch_size;
+    for (int batch = 0; batch < batch_size; ++batch) {
+      E* current_output = output + batch * size;
+      for (int i = 0; i < size; ++i) {
+        int rev = 0;
+        for (int j = 0; j < logn; ++j) {
+          if (i & (1 << j)) { rev |= 1 << (logn - 1 - j); }
+        }
+        if (i < rev) { std::swap(current_output[i], current_output[rev]); }
+      }
+    }
+    return eIcicleError::SUCCESS;
+  }
+
+  template <typename S = scalar_t, typename E = scalar_t>
+  eIcicleError cpu_ntt_ref(const Device& device, const E* input, int size, NTTDir dir, NTTConfig<S>& config, E* output)
+  {
+    if (size & (size - 1)) { return eIcicleError::INVALID_ARGUMENT; }
+
+    int total_size = size * config.batch_size;
+    E* temp_elements = new E[total_size];
+
+    // ICICLE_LOG_DEBUG << "INITIAL INPUT";
+    // for (int i = 0; i < total_size; ++i) {
+    //     ICICLE_LOG_DEBUG << "input[" << i << "]: " << input[i];
+    // }
+
+    if (config.columns_batch) {
+      // Distribute input into columns if columns_batch is set
+      for (int col = 0; col < config.batch_size; ++col) {
+        for (int row = 0; row < size; ++row) {
+          temp_elements[col * size + row] = input[row * config.batch_size + col];
+        }
+      }
+    } else {
+      std::copy(input, input + total_size, temp_elements);
```

Comment:
you can make temp_elements point to input rather than copy which is expensive

Reply:
fixed

Reply:
Changed back to copy to ensure that freeing temp_elements does not cause the input to be freed, preventing a potential double-free error. 
added TODO: Later, remove temp_elements and perform all calculations in-place (implement NTT for the case where columns_batch=true, in-place).